tests/xoto3/dynamodb/write_versioned/example_unit_test.py

```diff
@@ -0,0 +1,76 @@
+"""Imagine you had the following business logic and you wanted to test
+it. Here's how you might go about that, using the
+xoto3.dynamodb.write_versioned API itself to set everything up and
+prove that it worked.
+
+The example here is the same as the code we have in the README.
+"""
+import pytest
+
+import xoto3.dynamodb.write_versioned as wv
+
+# threadsafe because they are pure values.
+task_table = wv.ItemTable("TaskTable", item_name="Task")
+user_table = wv.ItemTable("UserTable", item_name="User")
+
+# imagine these were passed in via an API
+task_key = dict(id="task1")
+new_task = dict(task_key, name="plant tree", num_subtasks=3, user_id="steve")
+
+# this transaction builder is a pure function that constructs
+# a value representing DynamoDB effects to be transactionally/atomically applied.
+def create_and_link_task_unless_exists(t: wv.VersionedTransaction) -> wv.VersionedTransaction:
+    t = task_table.presume(task_key, None)(t)
+    # ^ we presume that the task does not exist, i.e. has the value None
+    # This is a no-op if a value has already been fetched.
+    # This is also purely an optimization to avoid an initial read;
+    # the whole transaction would result in the same data in the table without it.
+
+    existing_task = task_table.get(task_key)(t)
+    if existing_task:
+        return t
+        # ^ perform no action as long as the item already exists
+
+    t = task_table.put(new_task)(t)
+    # ^ put this item into this table as long as the item does not exist
+    user = user_table.require(dict(id=new_task["user_id"]))(t)  # type: ignore
+    # ^ fetch the user and raise an exception if it does not exist
+    user["task_ids"].append(task_key["id"])
+    t = user_table.put(user)(t)
+    # ^ make sure that the user knows about its task
+    return t
+    # ^ return the built transaction to be executed.
+
+
+def test_task_and_user_are_not_written_if_task_already_exists():
+    t = wv.VersionedTransaction(dict())
+    existing_task = dict(task_key, name="water flowers", is_done=True, user_id="felicity")
+    t = task_table.presume(task_key, existing_task)(t)
+    # presume lets us set up a fixture in the fake table, represented by the VersionedTransaction
+
+    t = create_and_link_task_unless_exists(t)
+
+    resulting_task = task_table.require(task_key)(t)
+    assert resulting_task == existing_task
+    # the task remains unchanged
+
+    with pytest.raises(wv.ItemNotFetchedException):
+        user_table.get(dict(id="felicity"))(t)
```

Comment:
I think I get it now. This is because it is `versioned_transact_write_items` that actually does the commit, and ultimately results in `None` being returned. The fact that the builder results in this exception is trapped there and turned into `None`. Is that correct?

Reply:
no. this doesn't return None because we never fetched this value at all - we have no idea what it is. The writer of the test didn't set that up as part of the fixture, and didn't provide any alternative way of discovering it.

At runtime within `v_t_w_i`, you would never see this because we catch this exception and fetch the item you asked for from an actual table.

Let me know if the docs on this exception need to be further clarified.

Reply:
i added some further docs and a new name. hopefully they'll help clarify what this Exception means